pkg/jsonmerge/object.go

```diff
@@ -0,0 +1,73 @@
+package jsonmerge
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+
+	"dario.cat/mergo"
+)
+
+type objectMerger struct {
+	io.Writer
+	dst map[string]interface{}
+}
+
+// NewObjectMerger creates a Merger for JSON objects.
+func NewObjectMerger(w io.Writer) Merger {
+	return &objectMerger{
+		Writer: w,
+		dst:    make(map[string]interface{}),
+	}
+}
+
+func (merger *objectMerger) NewPage(r io.Reader, isLastPage bool) io.ReadCloser {
+	return &objectMergerPage{
+		merger: merger,
+		Reader: r,
+	}
+}
+
+func (merger *objectMerger) Close() error {
+	// Marshal to JSON and write to output.
+	buf, err := json.Marshal(merger.dst)
+	if err != nil {
+		return err
+	}
+
+	_, err = merger.Writer.Write(buf)
+	return err
+}
+
+type objectMergerPage struct {
+	merger *objectMerger
+
+	io.Reader
+	buffer bytes.Buffer
+}
+
+func (page *objectMergerPage) Read(p []byte) (int, error) {
+	// Read into a temporary buffer to be merged and written later.
+	p = make([]byte, len(p))
+	n, err := page.Reader.Read(p)
+	if err != nil {
+		return 0, err
+	}
+	if n == 0 {
```

Comment:
I'm not sure about this conditional. I believe that a reader should only be returning (`0, nil`) when `len(p) == 0` as described in the `io.Reader` interface docs. However, with this conditional, in the case of `len(p) == 0` we're saying "there's no more to read" which isn't strictly correct.

Since we only seem to use this in `io.Copy` I'm not sure that we'll ever fall afoul of it but it also seems like something that might cause confusion?

Reply:
`io.Copy` never returns if we don't return `io.EOF`. It'll keep spinning. Tests were timing out, and jumping into the source of `io.Copy` showed why. But given this is effectively redirecting the response to a cache, telling any caller that nothing was read seems appropriate. The caller wouldn't have anything to read and the `io.Reader` truly would be at `io.EOF`.

Reply:
I can't make any test in the cli or `go-gh` fail without this line and I don't really see why it would. There's definitely no test  in `go-gh` which hits this code path. Can you point to the bit that fails without this?

Here's a test that demonstrates what I mean:

```go
func TestObjectMerger_ReadZeroThenSome(t *testing.T) {
	w := &bytes.Buffer{}
	merger := NewObjectMerger(w)

	r1 := bytes.NewBufferString(`{"a":1,"b":2}`)
	p1 := merger.NewPage(r1, true)

	var buf []byte
	for i := 0; i < 10; i++ {
		buf = make([]byte, i)
		_, err := p1.Read(buf)
		if err == io.EOF {
			break
		}
	}

	require.NoError(t, merger.Close())
	assert.JSONEq(t, `{"a":1,"b":2}`, w.String())
}
```

Functionally this is a `Copy` but starting from a byte slice size of `0`. Since `len(p) == 0`, the `BufferString.Read` function dutifully returns `0` as described in the `io.Reader` interface doc comment but we then treat that as `io.EOF` even though there is in fact the entire JSON string left to read.

In practice I don't know when this is an issue but given that it's explicitly called out in the `io.Reader` doc comment it makes me nervous doing this, especially since I don't understand what it's actually solving.

Reply:
Resolved with change to `io.CopyN`.